### PR TITLE
Bernstein polynomial flow extrapolation

### PR DESCRIPTION
**Pull request by MArpogaus**
This adresses #36.

### CODE REVIEW

**Comment by MArpogaus** [comment]
Just mentioning @oduerr here to notify him of the PR.

File: zuko/transforms.py [hunk 1]
```diff
@@ -629,42 +629,112 @@ class BernsteinTransform(MonotonicTransform):
...
+
+        theta_min = unconstrained_theta[..., :1]
+        unconstrained_theta = unconstrained_theta[..., 1:]
+
+        if smooth_bounds:
+            unconstrained_theta = torch.cat((
+                unconstrained_theta[..., :1],
+                unconstrained_theta,
+                unconstrained_theta[..., -1:],
+            ))
```
Comment (francois-rozet) [inline zuko/transforms.py:669]:
```suggestion
            unconstrained_theta = torch.cat((
                unconstrained_theta[..., :1],
                unconstrained_theta,
                unconstrained_theta[..., -1:],
            ), dim=-1)
```

File: zuko/transforms.py [hunk 2]
```diff
@@ -629,42 +629,112 @@ class BernsteinTransform(MonotonicTransform):
...
+        widths = torch.nn.functional.softplus(unconstrained_theta)
+
+        widths = torch.cat((theta_min, widths), dim=-1)
```
Comment (francois-rozet) [inline zuko/transforms.py:672]:
```suggestion

        widths = torch.nn.functional.softplus(unconstrained_theta)
        widths = torch.cat((theta_min, widths), dim=-1)
```

File: zuko/transforms.py [hunk 3]
```diff
@@ -629,42 +629,112 @@ class BernsteinTransform(MonotonicTransform):
...
+            self.order - 1, device=theta.device, dtype=theta.dtype
+        )
+
+        # save slope on boundaries for interpolation
+        x = torch.tensor([self.eps, 1 - self.eps], device=theta.device, dtype=theta.dtype)
+        rank = self.theta.dim()
+        if rank > 1:
+            # add singleton batch dimensions
+            dims = [...] + [None] * (rank - 1)
+            x = x[dims]
```
Comment (francois-rozet) [inline zuko/transforms.py:651]: This should not be necessary as `torch.distributions.Beta.log_prob` broadcasts.

File: zuko/transforms.py [hunk 4]
```diff
@@ -629,42 +629,112 @@ class BernsteinTransform(MonotonicTransform):
...
 
-        self.theta = self._increasing(theta)
-        self.linear = linear
+        self.theta = self._increasing(theta, smooth_bounds)
 
-        degree = theta.shape[-1]
-        alpha = torch.arange(1, degree + 1, device=theta.device, dtype=theta.dtype)
-        beta = torch.arange(degree, 0, -1, device=theta.device, dtype=theta.dtype)
+        self.order = self.theta.shape[-1] - 1
+        self.dtheta = self.order * (self.theta[..., 1:] - self.theta[..., :-1])
```
Comment (francois-rozet) [inline zuko/transforms.py:638]: It is not necessary to store `dtheta` and `dbasis` anymore.

File: zuko/transforms.py [hunk 5]
```diff
@@ -629,42 +629,112 @@ class BernsteinTransform(MonotonicTransform):
...
+        widths = torch.cat((theta_min, widths), dim=-1)
+
+        theta = torch.cumsum(widths, dim=-1) - shift
+
+        return theta
 
-        widths = torch.nn.functional.softplus(theta[..., 1:])
-        widths = torch.cat((theta[..., :1], widths), dim=-1)
+    @staticmethod
+    def get_bernstein_basis(order, **kwds):
```
Comment (francois-rozet) [inline zuko/transforms.py:679]: `_bernstein_basis`. The `_` is to prevent Sphinx from documenting the method. Also use `kwargs` instead of `kwds` and use type hints.

File: zuko/flows/polynomial.py [hunk 6]
```diff
@@ -87,13 +87,13 @@ def __init__(
         features: int,
         context: int = 0,
         degree: int = 16,
-        linear: bool = False,
+        bound: float = 10.0,
```
Comment (francois-rozet) [inline zuko/flows/polynomial.py:90]: Providing the `bound` option goes against the current convention to not provide transformation options in the flow constructors. Users are expected to build their own flows if they want more control. I am happy to reconsider this though, as it gives more flexibility, but not in the current PR as it would require changes in other places in the code base.

File: zuko/transforms.py [hunk 7]
```diff
@@ -605,10 +605,11 @@ class BernsteinTransform(MonotonicTransform):
 
     .. math:: f(x) = \frac{1}{M + 1} \sum_{i=0}^{M} b_{i+1,M-i+1}(\sigma(x)) \, \theta_i
 
-    where :math:`b_{i,j}` are the Bernstein basis polynomials and :math:`\sigma(x)` is
-    the sigmoid function.
+    where :math:`b_{i,j}` are the Bernstein basis polynomials and :math:`\sigma(x)` is the sigmoid function.
 
     References:
+        | Deep transformation models: Tackling complex regression problems with neural network based transformation models (Sick et al., 2020)
+        | https://arxiv.org/abs/2004.00464
```
Comment (francois-rozet) [inline zuko/transforms.py:612]:
```suggestion
        | https://arxiv.org/abs/2004.00464
        
```

File: zuko/transforms.py [hunk 8]
```diff
@@ -605,10 +605,11 @@ class BernsteinTransform(MonotonicTransform):
 
     .. math:: f(x) = \frac{1}{M + 1} \sum_{i=0}^{M} b_{i+1,M-i+1}(\sigma(x)) \, \theta_i
 
-    where :math:`b_{i,j}` are the Bernstein basis polynomials and :math:`\sigma(x)` is
-    the sigmoid function.
+    where :math:`b_{i,j}` are the Bernstein basis polynomials and :math:`\sigma(x)` is the sigmoid function.
```
Comment (francois-rozet) [inline zuko/transforms.py:608]: This description is not accurate anymore.

File: zuko/transforms.py [hunk 9]
```diff
@@ -629,42 +629,112 @@ class BernsteinTransform(MonotonicTransform):
...
     def f(self, x: Tensor) -> Tensor:
-        if self.linear:
-            x = (x + self.bound) / (2 * self.bound)  # map [-B, B] to [0, 1]
-        else:
-            x = torch.nn.functional.sigmoid(x)  # map [-inf, inf] to [0, 1]
+        x = self.scale_data(x)
+
+        left_bound = x <= self.eps
+        right_bound = x >= 1 - self.eps
+        x_safe = torch.where(left_bound | right_bound, 0.5, x)
```
Comment (francois-rozet) [inline zuko/transforms.py:711]: I am afraid that the `0.5` might not work with all supported versions of PyTorch. To be tested. Also, `left_bound` and `right_bound` are not great names as they are out-of-bound indicators.

File: zuko/transforms.py [hunk 10]
```diff
@@ -629,42 +629,112 @@ class BernsteinTransform(MonotonicTransform):
...
+        basis = torch.distributions.Beta(alpha, beta)
+        return basis
+
+    @staticmethod
+    def b_poly(x, theta, basis):
+        b = basis.log_prob(x.unsqueeze(-1)).exp()
+        y_poly = torch.mean(b * theta, dim=-1)
+        return y_poly
+
+    def scale_data(self, x):
```
Comment (francois-rozet) [inline zuko/transforms.py:691]: Unnecessary few-line function used once. Fold this into `f`. Same for `unscale_data`, `_extrapolation` and `_inverse_extrapolation`.

File: README.md [hunk 11]
```diff
@@ -90,7 +90,7 @@ For more information, check out the documentation and tutorials at [zuko.readthe
 | `UNAF`  | 2019 | [Unconstrained Monotonic Neural Networks](https://arxiv.org/abs/1908.05164) |
 | `CNF`   | 2018 | [Neural Ordinary Differential Equations](https://arxiv.org/abs/1806.07366) |
 | `GF`    | 2020 | [Gaussianization Flows](https://arxiv.org/abs/2003.01941) |
-| `BPF`   | 2020 | [Bernstein-Polynomial Normalizing Flows](https://arxiv.org/abs/2204.13939) |
+| `BPF`   | 2020 | [Bernstein-Polynomial Normalizing Flows](https://arxiv.org/abs/2004.00464) |
```
Comment (francois-rozet) [inline README.md:93]: Should we cite to Sick et al. (2020) in the transformation and flow as well? I cited Arpogaus et al. (2022) as it was the first to use the term "Berstein-Polynomial Normalizing Flows".

Comment (oduerr) [reply README.md:93]: We should cite Sick et al. (2020) here as well since this was the first time Bernstein Polynomials have been used with NN to control a 1D flow.

File: zuko/flows/polynomial.py (hunk 6 above)
Comment (MArpogaus) [reply zuko/flows/polynomial.py:90]: ok, i'll revert it.

File: zuko/transforms.py (hunk 9 above)
Comment (MArpogaus) [reply zuko/transforms.py:711]: Would this be better:

            x_safe = torch.where(lower_bound | upper_bound, 0.5 * torch.ones_like(x), x)

File: README.md (hunk 11 above)
Comment (MArpogaus) [reply README.md:93]: Yes, that's why I changed it to Sick at al. here, but kept both references in the doc string.

File: zuko/transforms.py (hunk 3 above)
Comment (MArpogaus) [reply zuko/transforms.py:651]: Test fail if I remove this. Couldn't make it work without.
What would you suggest?

File: zuko/transforms.py [hunk 12]
```diff
@@ -633,17 +633,12 @@ class BernsteinTransform(MonotonicTransform):
...
-            theta, smooth_bounds, self.bound if keep_in_bounds else None
-        )
-        self.keep_in_bounds = keep_in_bounds
-
+        self.theta = self._constrain_theta(theta)
         self.order = self.theta.shape[-1] - 1
+
         dtheta = self.order * (self.theta[..., 1:] - self.theta[..., :-1])
 
         self.basis = self._bernstein_basis(self.order, device=theta.device, dtype=theta.dtype)
```
Comment (francois-rozet) [inline zuko/transforms.py:644]: Is this part necessary for the bounded transform? I guess that offsets should be $-B$ and $B$ and that slopes should be $1$? Or maybe there are numerical errors?

Also if you don't want to expand the dimensions of `x`, you can evaluate the Bernstein polynomial(s) at `self.eps` and `1 - self.eps` separately. Scalars always broadcast!

File: zuko/transforms.py [hunk 13]
```diff
@@ -30,8 +31,7 @@
 
 from textwrap import indent
 from torch import BoolTensor, LongTensor, Size, Tensor
-from torch.distributions import Transform, constraints, Distribution
-from torch.distributions.transforms import *  # noqa: F403
```
Comment (francois-rozet) [inline zuko/transforms.py:34]: Keep this line. It allows users to import any torch transform from `zuko.transforms`, which is very convenient. This is not a safety hazard because F405 (undefined variable or defined from star imports) is not allowed within our code base.

File: zuko/transforms.py [hunk 14]
```diff
@@ -633,17 +633,12 @@ class BernsteinTransform(MonotonicTransform):
     bijective = True
     sign = +1
 
-    def __init__(
-        self, theta: Tensor, smooth_bounds: bool = True, keep_in_bounds: bool = False, **kwargs
-    ):
+    def __init__(self, theta: Tensor, **kwargs):
```
Comment (francois-rozet) [inline zuko/transforms.py:636]: Update the docstring.

File: zuko/transforms.py [hunk 15]
```diff
@@ -661,34 +656,27 @@ def __init__(
...
-            unconstrained_theta = unconstrained_theta[..., 1:]
-            fn = torch.nn.functional.softplus
+    def _constrain_theta(unconstrained_theta: Tensor) -> Tensor:
+        """Processes the unconstrained output of the hyper-network to be increasing."""
+        shift = math.log(2.0) * unconstrained_theta.shape[-1] / 2
+
+        theta_min = unconstrained_theta[..., :1]
+        unconstrained_theta = unconstrained_theta[..., 1:]
+
+        # ensure smooth bounds
```
Comment (francois-rozet) [inline zuko/transforms.py:666]: Did you derive the math of these yourself or is it based on some reference? If the latter it might be worth linking to it. Same for the bounded transform actually.

File: zuko/transforms.py [hunk 16]
```diff
@@ -743,6 +731,34 @@ def _inverse(self, y: Tensor):
         return x
 
 
+class BoundedBernsteinTransform(BernsteinTransform):
```
Comment (francois-rozet) [inline zuko/transforms.py:734]: Add a docstring. No need for a long description (unless you want to explain the math), but say that the $[-B, B]$ domain interval is mapped to the $[-B, B]$ co-domain interval.

**Review by francois-rozet — COMMENTED** [review]
Great work and very clean code! I just have a few minor comments.

File: zuko/flows/polynomial.py [hunk 17]
```diff
@@ -1,16 +1,16 @@
 r"""Polynomial flows."""
 
 __all__ = [
-    'SOSPF',
-    'BPF',
+    "SOSPF",
+    "BPF",
```
Comment (francois-rozet) [inline zuko/flows/polynomial.py:5]:
```suggestion
    'SOSPF',
    'BPF',
```

File: zuko/flows/polynomial.py [hunk 18]
```diff
@@ -86,18 +85,12 @@ def __init__(
...
         context: int = 0,
         degree: int = 16,
-        linear: bool = False,
         **kwargs,
     ):
         super().__init__(
             features=features,
             context=context,
-            univariate=partial(BernsteinTransform, linear=linear),
+            univariate=partial(BoundedBernsteinTransform),
```
Comment (francois-rozet) [inline zuko/flows/polynomial.py:93]:
```suggestion
            univariate=BoundedBernsteinTransform,
```

File: zuko/transforms.py (hunk 12 above)
Comment (MArpogaus) [reply zuko/transforms.py:640]: Good point. I refactored the calculations into a method and overload it with one returning fixed values in the bounded subclass.

File: zuko/transforms.py (hunk 13 above)
Comment (MArpogaus) [reply zuko/transforms.py:34]: Sorry, was not intended. Reverted it

File: zuko/transforms.py (hunk 15 above)
Comment (MArpogaus) [reply zuko/transforms.py:673]: I think this was initially @oduerr's idea.

He observed, that the small initial values for thetas lead to a rather sub optimal initialization and hence decided to add this offset to distribute the symmetrically around the normal base distribution.

$\log(2)$ is used because $\text{softplus}(0) = \log(2)$

**Review by MArpogaus — COMMENTED** [review]
I think i resolved all open issues. Anything else that holds us back from closing this PR?

**Comment by francois-rozet** [comment]
Hello @MArpogaus, I built the documentation with the new components and there was a few errors, so I made a few edits + some reformulation. I also limited the domain to [-5, 5] (like NSF) instead of [-10, 10], which should improve the results of `BPF` for standardized (zero mean, unit variance) data.

Otherwise everything seems fine to me. Great job! I am ready to merge the PR, so tell me if you agree with my edits.

File: zuko/transforms.py [hunk 19]
```diff
@@ -739,34 +742,33 @@ def _inverse(self, y: Tensor) -> Tensor:
...
+    that the polynomial's domain and codomain match the interval :math:`[-B, B]`. It
+    also enforces that the first derivative at the bounds is one and that the second
+    derivative is zero, ensuring a smooth transition to the identity function outside
+    the bounds.
+
+    These constraints make the transformation suitable for chaining in flows.
+
+    Arguments:
+        theta: The unconstrained polynomial coefficients :math:`\theta`,
+            with shape :math:`(*, M - 5)`.
```
Comment (francois-rozet) [inline zuko/transforms.py:757]: To be sure, is this correct @MArpogaus?
